clang/test/SemaSYCL/fpga_pipes.cpp

```diff
@@ -33,11 +35,51 @@ int Storage4 __attribute__((io_pipe_id(5)));
 template <int N>
 pipe_storage Storage5 __attribute__((io_pipe_id(N)));
 
+// Test that checks wrong function template instantiation and ensures that the type
+// is checked properly when instantiating from the template definition.
+template <typename Ty>
+// expected-error@+1 {{integral constant expression must have integral or unscoped enumeration type, not 'S'}}
+const pipe_storage Storage6 __attribute__((io_pipe_id(Ty{})));
+
+struct S {};
+
+// Test that checks template instantiations for different arg values.
+template <int A>
+pipe_storage Storage7 __attribute__((io_pipe_id(1))) // expected-note {{previous attribute is here}}
+__attribute__((io_pipe_id(A)));                      // expected-warning{{attribute 'io_pipe_id' is already applied with different arguments}}
```

Comment:
I'd suggest to error out in case of redeclaration appears.

It's not expected that a common user would use this attribute. Instead it's only used on a static member (pipe storage struct) of pipe class within SYCL headers and it accepts a constexpr integer coming from pipe class template parameter as the attribute's argument. If someone modifies SYCL headers in a way of redeclaring the structure (somehow) - they should get a strong recommendation from the compiler not to do that.

Back then there was no better alternative other than an attribute. May be now we can consider removing this attribute and replacing it with implicit pipe property_list, but it goes out of a scope of this PR.

Reply:
@MrSidims, I have added error in case of redeclaration happens. Thanks for the review.

>>Back then there was no better alternative other than an attribute. May be now we can consider removing this attribute and replacing it with implicit pipe property_list, but it goes out of a scope of this PR.

Agreed. This will be of out a scope of this PR.